gemrb/core/Scriptable/Scriptable.cpp

```diff
@@ -2174,51 +2167,152 @@ unsigned char Movable::GetNextFace()
 	return Orientation;
 }
 
-// returns whether we made all pending steps (so, false if we must be called again this tick)
-// we can't just do them all here because the caller might have to update searchmap etc
-bool Movable::DoStep(unsigned int walk_speed, ieDword time)
+
+void Movable::Backoff()
+{
+	StanceID = IE_ANI_READY;
+	if (InternalFlags & IF_RUNNING) {
+		randomBackoff = RAND(MAX_PATH_TRIES * 2 / 3, MAX_PATH_TRIES * 4 / 3);
+	} else {
+		randomBackoff = RAND(MAX_PATH_TRIES, MAX_PATH_TRIES * 2);
+	}
+}
+
+
+void Movable::BumpAway()
 {
+	area->ClearSearchMapFor(this);
+	if (!IsBumped()) oldPos = Pos;
+	bumped = true;
+	bumpBackTries = 0;
+	area->AdjustPositionNavmap(Pos);
+}
+
+// Takes care of movement and actor bumping, i.e. gently pushing blocking actors out of the way
+// The movement logic is a proportional regulator: the displacement/movement vector has a
+// fixed radius, based on actor walk speed, and its direction heads towards the next waypoint.
+// The bumping logic checks if there would be a collision if the actor was to move according to this
+// displacement vector and then, if that is the case, checks if that actor can be bumped
+// In that case, it bumps it and goes on with its step, otherwise it either stops and waits
+// for a random time (inspired by network media access control algorithms) or just stops if
+// the goal is close enough.
+void Movable::DoStep(unsigned int walkScale, ieDword time) {
+	// Only bump back if not moving
+	// Actors can be bumped while moving if they are backing off
 	if (!path) {
-		return true;
+		if (!IsBumped()) {
+			return;
+		}
+		area->ClearSearchMapFor(this);
+		unsigned oldPosBlockStatus = area->GetBlockedNavmap(oldPos.x, oldPos.y);
+		if (!(oldPosBlockStatus & PATH_MAP_PASSABLE)) {
+			// Do bump back if the actor is "blocking" itself
+			if (!(oldPosBlockStatus & PATH_MAP_ACTOR && Type == ST_ACTOR && area->GetActor(oldPos, GA_NO_DEAD|GA_NO_UNSCHEDULED) == (Actor*)this)) {
+				area->BlockSearchMap(Pos, size, ((Actor*)this)->IsPartyMember() ? PATH_MAP_PC : PATH_MAP_NPC);
+				Log(DEBUG, "DoStep", "%s not bumping back, blocked by %d", GetName(0), oldPosBlockStatus);
+				if (Type == ST_ACTOR && ((Actor*)this)->GetStat(IE_EA) < EA_GOODCUTOFF) {
+					bumpBackTries++;
+					if (bumpBackTries > MAX_BUMP_BACK_TRIES) {
+						if (SquaredDistance(Pos, oldPos) < unsigned(size * 32 * size * 32)) {
+							oldPos = Pos;
+							bumped = false;
+							bumpBackTries = 0;
+							if (SquaredDistance(Pos, Destination) < unsigned(size * 32 * size * 32)) {
+								ClearPath(true);
+							}
+						}
+					}
+				}
+
+				return;
+			}
```

Comment:
This ends up with 7 levels of nesting. Perhaps split it out for better readability? That would take out 2 levels, then you can merge the MAX_BUMP_BACK_TRIES one with the next and it's immediately more palatable. Sonar has a special check for this.
I guess also the initial check could be inverted to exit early, but I don't know if we need to go that far.